provider/query_parameter_value.go

```diff
@@ -0,0 +1,52 @@
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Query parameter holds normalized parameter data ready to be inserted in the
+//  final query
+type QueryParameterValue struct {
+	// Token to replace e.g., !TOKEN!
+	Token string
+	// SQL expression to be inserted. Contains "?" that will be replaced with an
+	//  ordinal argument e.g., "$1"
+	SQL string
+	// Value that will be passed to the final query in arguments list
+	Value interface{}
+	// Raw parameter values for debugging and monitoring
+	RawValues map[string]string
+}
+
+// ReplaceParams substitutes configured query parameter tokens for their values
+//  within the provided SQL string
+func ReplaceParams(params map[string]QueryParameterValue, sql string, args *[]interface{}) string {
+	if params == nil {
+		return sql
+	}
+
+	for _, token := range ParameterTokenRegexp.FindAllString(sql, -1) {
+		param := params[token]
+
+		// Replace every ? in the param's SQL with a positional argument
+		paramSQL := ""
+		argFound := false
+		for _, c := range param.SQL {
+			if c == '?' {
+				if !argFound {
+					*args = append(*args, param.Value)
+					argFound = true
+				}
+				paramSQL += fmt.Sprintf("$%d", len(*args))
+			} else {
+				paramSQL += string(c)
+			}
+		}
+
+		// Finally, replace current token with the prepared SQL
+		sql = strings.Replace(sql, token, paramSQL, 1)
```

Comment:
I'd suggest:
```suggestion
		sql = strings.ReplaceAll(sql, token, paramSQL)
```
And in the beginning of the loop check if a token has already been encountered. This way we won't have to pass the same arg multiple times if the token is used more than once.